consensus/service/src/client_api_service.rs

```diff
@@ -26,135 +25,468 @@ use std::sync::Arc;
+pub struct ClientApiService {
+    enclave: Arc<dyn ConsensusEnclave + Send + Sync>,
+    tx_manager: Arc<dyn TxManager + Send + Sync>,
+    ledger: Arc<dyn Ledger + Send + Sync>,
+    /// Passes proposed transactions to the consensus service.
+    propose_tx_callback: ProposeTxCallback,
+    /// Returns true if this node is able to process proposed transactions.
     is_serving_fn: Arc<(dyn Fn() -> bool + Sync + Send)>,
     logger: Logger,
 }
 
-impl<E: ConsensusEnclaveProxy, L: Ledger + Clone> ClientApiService<E, L> {
+impl ClientApiService {
     pub fn new(
-        enclave: E,
+        enclave: Arc<dyn ConsensusEnclave + Send + Sync>,
         scp_client_value_sender: ProposeTxCallback,
-        ledger: L,
-        tx_manager: Arc<TxManagerImpl<E, DefaultTxManagerUntrustedInterfaces<LedgerDB>>>,
+        ledger: Arc<dyn Ledger + Send + Sync>,
+        tx_manager: Arc<dyn TxManager + Send + Sync>,
         is_serving_fn: Arc<(dyn Fn() -> bool + Sync + Send)>,
         logger: Logger,
     ) -> Self {
         Self {
             enclave,
-            scp_client_value_sender,
             tx_manager,
             ledger,
+            propose_tx_callback: scp_client_value_sender,
             is_serving_fn,
             logger,
         }
     }
 
-    fn real_client_tx_propose(
+    /// Handles a client's proposed transaction.
+    ///
+    /// # Arguments
+    /// `msg` - An encrypted message from a client to the enclave.
+    /// `logger` - Logger
+    fn handle_proposed_tx(
         &mut self,
-        request: Message,
-        logger: &Logger,
+        msg: Message,
     ) -> Result<ProposeTxResponse, ConsensusGrpcError> {
         counters::ADD_TX_INITIATED.inc();
 
-        if counters::CUR_NUM_PENDING_VALUES.get() > PENDING_LIMIT {
-            self.enclave.client_discard_message(request.into())?;
-
-            log::trace!(
-                logger,
-                "Ignoring add transaction call, node is over capacity."
-            );
-            return Err(ConsensusGrpcError::OverCapacity);
-        }
+        let tx_context = self.enclave.client_tx_propose(msg.into())?;
 
-        // Check if node is accepting requests.
-        if !(self.is_serving_fn)() {
-            self.enclave.client_discard_message(request.into())?;
+        let num_blocks = self.ledger.num_blocks().map_err(ConsensusGrpcError::from)?;
+        let mut response = ProposeTxResponse::new();
+        response.set_num_blocks(num_blocks);
 
-            log::info!(
-                logger,
-                "Ignoring add transaction call, not currently serving requests."
-            );
-            return Err(ConsensusGrpcError::NotServing);
+        // Reject the proposed transaction if it contains any key images that have already been spent.
+        // This is done here as a courtesy to give clients immediate feedback about their transaction.
+        if self.contains_spent_key_image(&tx_context)? {
```

Comment:
Checking for the tombstone block should also be helpful for clients.

Reply:
Good idea. I need to get this PR merged, so I'll do that in a follow-on PR.